packages/docs-infra/src/CodeHighlighter/loadFallbackCode.ts

```diff
@@ -0,0 +1,475 @@
+import type {
+  Code,
+  VariantExtraFiles,
+  ParseSource,
+  LoadSource,
+  LoadVariantMeta,
+  VariantSource,
+  LoadCodeMeta,
+  VariantCode,
+} from './types';
+import { loadVariant } from './loadVariant';
+import { getFileNameFromUrl } from '../pipeline/loaderUtils';
+
+// Helper function to get the source for a specific filename from a variant
+async function getFileSource(
+  variant: VariantCode,
+  requestedFilename: string | undefined,
+  loadSource: LoadSource | undefined,
+): Promise<{ source: VariantSource; filename: string | undefined }> {
+  const filename = requestedFilename || variant.fileName;
+
+  if (!filename) {
+    // If no filename is available, return the main variant source
+    if (variant.source !== undefined) {
+      return { source: variant.source, filename: undefined };
+    }
+    throw new Error('No filename available and no source in variant');
+  }
+
+  // If requesting the main file and we have its source
+  if (filename === variant.fileName && variant.source !== undefined) {
+    return { source: variant.source, filename };
+  }
+
+  // If requesting an extra file and we have its source
+  if (filename !== variant.fileName && variant.extraFiles) {
+    const extraFile = variant.extraFiles[filename];
+    if (extraFile && typeof extraFile !== 'string' && extraFile.source !== undefined) {
+      return { source: extraFile.source, filename };
+    }
+
+    // If we have the URL but not the source, we need to load it
+    if (typeof extraFile === 'string' && loadSource) {
+      const loadResult = await loadSource(extraFile);
+      return { source: loadResult.source, filename };
+    }
+
+    if (extraFile && typeof extraFile !== 'string' && !extraFile.source && loadSource) {
+      // This case shouldn't normally happen, but handle it anyway
+      throw new Error(`Extra file ${filename} has no source or URL to load from`);
+    }
+  }
+
+  throw new Error(`File ${filename} not found in variant or cannot be loaded`);
+}
+
+export type FallbackVariants = {
+  code: Code;
+  initialFilename: string | undefined;
+  initialSource: VariantSource;
+  initialExtraFiles?: VariantExtraFiles;
+  allFileNames: string[];
+  processedGlobalsCode?: Array<Code>;
+};
+
+export async function loadFallbackCode(
```

Comment:
This function has too many attributes, and most of them optional, can we convert it to an options object, or maybe have the required inputs (url, initialVariant, loaded as params and then follow them with options input):
```
export async function loadFallbackCode(
  url: string,
  initialVariant: string,
  loaded: Code | undefined,
  options: LoadFallbackOptions
)
```